zebrad/src/components/mempool/crawler.rs

```diff
@@ -0,0 +1,88 @@
+//! Zebra Mempool crawler.
+//!
+//! The crawler periodically requests transactions from peers in order to populate the mempool.
+
+use std::time::Duration;
+
+use futures::{stream, StreamExt, TryStreamExt};
+use tokio::{sync::Mutex, time::sleep};
+use tower::{BoxError, Service, ServiceExt};
+
+use zebra_network::{Request, Response};
+
+/// The number of peers to request transactions from per crawl event.
+const FANOUT: usize = 4;
+
+/// The delay between crawl events.
+const RATE_LIMIT_DELAY: Duration = Duration::from_secs(75);
+
+/// The mempool transaction crawler.
+pub struct Crawler<S> {
+    peer_set: Mutex<S>,
+}
+
+impl<S> Crawler<S>
+where
+    S: Service<Request, Response = Response, Error = BoxError> + Clone + Send + 'static,
+    S::Future: Send,
+{
+    /// Spawn an asynchronous task to run the mempool crawler.
+    pub fn spawn(peer_set: S) {
+        let crawler = Crawler {
+            peer_set: Mutex::new(peer_set),
+        };
+
+        tokio::spawn(crawler.run());
+    }
+
+    /// Periodically crawl peers for transactions to include in the mempool.
+    pub async fn run(self) {
```

Comment:
`run` is almost identical to `ChainSync::sync`.
I wonder why we need a `Mutex` here, but not there?

https://github.com/ZcashFoundation/zebra/blob/6f8f4d8987d7fc1e416de6c1faf3adcfafd4f71f/zebrad/src/components/sync.rs#L283

Reply:
The problem is that `tokio::spawn()` requires the `Future` to be executed on a new task to be `Send`. `ChainSync.sync()` isn't `Send` either, but that's not an issue because it's just `.await`ed inside a `Future` (`StartCmd.start()`) that the `TokioComponent` eventually blocks on. I.e., for `RuntimeRun::run(&mut self, fut: impl Future<Output = Result<(), Report>>)` the `fut` does not have to be `Send`.

One interesting solution might be to do something like `futures::join!(syncer.sync(), crawler.run())`. That would remove the need for the `Mutex`, but the disadvantage would be that the crawler will always run on the same thread as the chain synchronizer. Not sure what's best here, I'm fine either way :grin: